### PR TITLE
Add flags module

### PR DESCRIPTION
**Pull request by hakonhall**
FileFlagSource reads flags from files in /etc/vespa/flags and is a component
that can be injected in host admin, config server, etc. A flag named foo
corresponds to filename foo.

In general a FlagSource manages:
 - Feature flags: A feature is either set (true/enabled) or otherwise false.
   Touching a file foo means the feature flag foo is set (true).
 - Value flags: Either a String or empty if not set. The String corresponds to
   the file content.

The plan is to make the config server another source of flags. A unified
FlagSource can merge the two sources with some priority and used in e.g. parts
of node-admin. In other parts one would only have access to the file source.

Defines various flag facades:
 - FeatureFlag: Used to test whether a feature has been enabled or not.
 - IntFlag
 - JacksonFlag: Deserializes JSON to Jackson class, or return default if unset.
 - LongFlag
 - OptionalJacksonFlag: Deserializes JSON to Jackson class, or empty if unset.
 - OptionalStringFlag
 - StringFlag

This is part of removing some of the last Chef recipes. Some minor tweaks have
been necessary as part of this and are included in this PR (test whether a
systemd service exists, task-friendly file deletion, allow capitalized letters
in YUM package name).

### CODE REVIEW

File: flags/src/main/java/com/yahoo/vespa/flags/JacksonFlag.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    public static <T> Function<FlagSource, JacksonFlag<T>> createUnbound(String flagId, Class<T> jacksonClass, T defaultValue) {
+        return createUnbound(new FlagId(flagId), jacksonClass, defaultValue);
+    }
+
+    public static <T> Function<FlagSource, JacksonFlag<T>> createUnbound(FlagId id, Class<T> jacksonClass, T defaultValue) {
+        return source -> new JacksonFlag<>(id, jacksonClass, defaultValue, source);
+    }
+
+    public JacksonFlag(String flagId, Class<T> jacksonClass, T defaultValue, FlagSource source) {
```
Comment (freva) [inline flags/src/main/java/com/yahoo/vespa/flags/JacksonFlag.java]: Can't we just get the class from the `defaultValue`?

**Review by freva — COMMENTED** [review]
I don't like the Optional classes.. Can't the regular classes both implement
`T value()` and `Optional<T> optionalValue()`? There would probably need to be 2 different constructors, with and without `defaultValue`. 

Is the `defaultValue` even needed in the constructor? Why not `T value(T defaultValue)`?

File: flags/src/main/java/com/yahoo/vespa/flags/JacksonFlag.java (hunk 1 above)
Comment (hakonhall) [reply flags/src/main/java/com/yahoo/vespa/flags/JacksonFlag.java]: I don't think so, e.g. defaultValue.getClass() returns the type Class<? extends Object>.

File: node-admin/vespa-node-admin.spec [hunk 2]
```diff
@@ -33,8 +33,9 @@ mkdir -p "$app_dir"/components
 cp node-admin/src/main/application/services.xml "$app_dir"
 
 declare -a jar_components=(
-  node-admin/target/node-admin-jar-with-dependencies.jar
   docker-api/target/docker-api-jar-with-dependencies.jar
+  flags/target/flags-jar-with-dependencies.jar
```
[the remark this answers is not in the document]
Comment (hakonhall) [reply node-admin/vespa-node-admin.spec]: It doesn't need flags, at least not yet

File: flags/src/main/java/com/yahoo/vespa/flags/FileFlagSource.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        return getBytes(id).map(bytes -> new String(bytes, StandardCharsets.UTF_8));
+    }
+
+    public Optional<byte[]> getBytes(FlagId id) {
+        try {
+            return Optional.of(Files.readAllBytes(getPath(id)));
+        } catch (NoSuchFileException e) {
+            return Optional.empty();
+        } catch (IOException e) {
+            throw new UncheckedTimeoutException(e);
```
[the remark this answers is not in the document]
Comment (hakonhall) [reply flags/src/main/java/com/yahoo/vespa/flags/FileFlagSource.java]: Nice catch, fixed.

**Comment by hakonhall** [comment]
On Optional classes:
 - We could do away with non-Optional classes if the default is specified where value is retrieved, as an .orElse(defaultValue) on the returned Optional is a simple API.
 - My hope is instead to encourage early-bind default values so different call sites doesn't have to reference the (presumably) same default instance. And I was thinking a default logically belongs to the definition of the flag, not to wherever and whenever it is used.
 - I don't see the use-case of both specifying a default but allow overriding it at retrieval time. That seems dangerous and confusing.
 - Sometimes a feature may be backed by additional parameters, in case an OptionalJacksonClass would be ideal. But if instead the flags provides tuning of already set values, a JacksonClass is preferable with a hard-coded defaultValue. Similar arguments holds for other types.
 - OptionalInt and OptionalLong can be implemented later if needed.

**Comment by hakonhall** [comment]
I just realized that optionality of a flag, like FlagSource::getString returning empty if feature is not specified by the source, is necessary when merging sources. Say there are 2 sources: file and config server, and say a file source has higher priority than config server. The final value of the flag is then
 - determined by the file value, if isPresent()
 - otherwise, determined by the config server, if isPresent()
 - otherwise, default.

This is impossible with hasFeature(). There is no way to know whether a file source has explicitly set a feature off, on, or not at all. In particular, if config server enables a feature, there's no way to disable it on file.

Therefore, let me remove the hasFeature() of ConfigSource and reimplement the FeatureFlag as true/false content of the flag file.

File: flags/src/main/java/com/yahoo/vespa/flags/JacksonFlag.java (hunk 1 above)
Comment (freva) [reply flags/src/main/java/com/yahoo/vespa/flags/JacksonFlag.java]: Can't it safely be casted to `Class<T>`?

**Comment by hakonhall** [comment]
To make the following equivalent:
 - A flag is not set in either sources
 - A flag is set in config server, but "unset" in file source.

The Optional flags classes has to be removed: The client should not be able to ask whether a flag has been set by the source. It can only ask what the value is.

File: flags/src/main/java/com/yahoo/vespa/flags/JacksonFlag.java (hunk 1 above)
Comment (hakonhall) [reply flags/src/main/java/com/yahoo/vespa/flags/JacksonFlag.java]: The best I could come up with is:
https://stackoverflow.com/questions/3437897/how-to-get-a-class-instance-of-generics-type-t

which recommends this type of API.

Comment (hakonhall) [reply flags/src/main/java/com/yahoo/vespa/flags/JacksonFlag.java]: https://stackoverflow.com/questions/36390184/how-to-get-class-instances-of-generic-types-in-generic-methods

recommends adding Class<T> parameter to a method taking 'T instance', in order to get an Class<T> instance.

Comment (freva) [reply flags/src/main/java/com/yahoo/vespa/flags/JacksonFlag.java]: Right, but in this case we do take in a default value which is an instance of `T`, so it should be safe to this:
```
 public JacksonFlag(String flagId, T defaultValue, FlagSource source) {
    Class<T> jacksonClass = (Class<T>) defaultValue.getClass();
    ...
```
?

File: flags/src/main/java/com/yahoo/vespa/flags/FeatureFlag.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        this(id, false, source);
+    }
+
+    private FeatureFlag(FlagId id, boolean defaultValue, FlagSource source) {
+        this.id = id;
+        this.defaultValue = defaultValue;
+        this.source = source;
+    }
+
+    public FeatureFlag defaultToTrue() {
```
Comment (freva) [inline flags/src/main/java/com/yahoo/vespa/flags/FeatureFlag.java]: This is just confusing, always force setting default value explicitly?

**Comment by hakonhall** [comment]
PTAL @freva

File: flags/src/main/java/com/yahoo/vespa/flags/FeatureFlag.java (hunk 4 above)
Comment (hakonhall) [reply flags/src/main/java/com/yahoo/vespa/flags/FeatureFlag.java]: Fixed

**Review by freva — APPROVED** [review]

**Pull request merged**